Sources/Sentry/SentrySessionTracker.m

```diff
@@ -82,28 +88,65 @@ - (void)stop
 #endif
 }
 
+/**
+ * End previously cached sessions. We never can be sure that WillResignActive or WillTerminate
+ are called due to a crash or unexpected behavior. Still, we don't want to lose such sessions and
+ end them.
+ */
+- (void)endCachedSession
+{
+    SentryHub *hub = [SentrySDK currentHub];
+    NSDate *_Nullable lastInForeground =
+        [[[hub getClient] fileManager] readTimestampLastInForeground];
+    if (nil != lastInForeground) {
+        [[[hub getClient] fileManager] deleteTimestampLastInForeground];
+    }
+
+    [hub closeCachedSessionWithTimestamp:lastInForeground];
+}
+
+/**
+ * Is only called when an app is receiving events / it is in the foreground.
+ */
 - (void)didBecomeActive
 {
-    NSDate *sessionEnded
-        = nil == self.lastInForeground ? [self.currentDateProvider date] : self.lastInForeground;
-    NSTimeInterval secondsInBackground =
-        [[self.currentDateProvider date] timeIntervalSinceDate:sessionEnded];
     SentryHub *hub = [SentrySDK currentHub];
-    if (secondsInBackground * 1000 > (double)(self.options.sessionTrackingIntervalMillis)) {
-        [hub endSessionWithTimestamp:sessionEnded];
+    self.lastInForeground = [[[hub getClient] fileManager] readTimestampLastInForeground];
+
+    if (nil == self.lastInForeground) {
+        // Cause we don't want to track sessions if the app is in the background we need to wait
+        // until the app is in the foreground to start a session.
         [hub startSession];
+    } else {
+        // When the app was already in the foreground we have to decide whether it was long enough
+        // in the background to start a new session or to keep the session open. We don't want a new
+        // session if the user switches to another app for just a few seconds.
+        NSTimeInterval secondsInBackground =
+            [[self.currentDateProvider date] timeIntervalSinceDate:self.lastInForeground];
+
+        if (secondsInBackground * 1000 >= (double)(self.options.sessionTrackingIntervalMillis)) {
+            [hub endSessionWithTimestamp:self.lastInForeground];
+            [hub startSession];
+        }
     }
-    [[[hub getClient] fileManager] deleteTimestampLastInForeground];
-    self.lastInForeground = nil;
```

Comment:
L: Clearing up this state no longer needed?

Reply:
`lastInForeground` is set in `willResignActive`. When the app moves back to foreground`didBecomeActive` is called. Now `lastInForeground` is invalid. It is going to be valid again when `willResignActive` is called. The only reason that should happen is when the app crashes. Ending the session then anyway ignores the `lastInForeground`, but `lastInForeground` doesn't tell the truth. If `willResignActive` and `willTerminate` are not called then we would end the session as a healthy one. If we would clear up the state as it was before it would be an abnormal session because `lastInForeground` is `nil`. 
The advantage of the solution right now is that we save one access to the disk.

Reply:
If we delete the `lastInForeground` in `didBecomeActive` I think we are not really accurate, because we know that the app was inForeground a while ago. We could set `lastInForeground` both in `willResignActive` and `willResignActive`.

Reply:
I added the cleaning up again, because we would end sessions as healthy if `willResignActive` and `willTerminate` are not called. Instead, they should be ended as abnormal. `testKillAppWithoutNotificationsAndNoCrash_EndsWithAbnormalSession` checks this edge case.